### ISSUE TITLE
DSL.currentTimestamp() cannot be used as a default value for a column, via alterTable

### ISSUE THREAD

**Issue by apatrida** [jOOQ/jOOQ#5140]
When `SQLDialect` is properly configured, it is ignored during this alter table causing a stack trace:

```
            db.alterTable("Accounts")
                    .alterColumn("dtCreated").defaultValue(DSL.currentTimestamp())
                    .execute()
```

and the error:

```
org.jooq.exception.SQLDialectNotSupportedException: Type class org.jooq.impl.CurrentTimestamp is not supported in dialect DEFAULT

	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:786)
	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:730)
	at org.jooq.impl.DSL.getDataType(DSL.java:14357)
	at org.jooq.impl.DSL.val(DSL.java:12745)
	at org.jooq.impl.DSL.inline(DSL.java:12603)
	at org.jooq.impl.AlterTableImpl.defaultValue(AlterTableImpl.java:170)
	at org.jooq.impl.AlterTableImpl.defaultValue(AlterTableImpl.java:78)
```

the call to `inline`  is static and does not retain the dialect currently set.   Therefore it calls `getDataType` which does not accept a dialect, although its docs says it is dialect specific (but it isn't):

```
   /**
     * Get the default data type for the {@link DSLContext}'s underlying
     * {@link SQLDialect} and a given Java type.
     *
     * @param <T> The generic type
     * @param type The Java type
     * @return The <code>DSL</code>'s underlying default data type.
     */
    @Support
    public static <T> DataType<T> getDataType(Class<T> type) {
        return DefaultDataType.getDataType(SQLDialect.DEFAULT, type);
    }
```

`getDataType` for `SQLDialect.DEFAULT` does not work for `CurrentTimeStamp` class and throws the exception reported above.  Therefore it is impossible this could be used in many places.

that appears to be a bug.

**Comment by lukaseder** [comment]
Thanks a lot for reporting!

Hmm, interesting. I suspect this is related to this weird Java language "regression" when generics were introduced: http://stackoverflow.com/q/5361513/521799. I can reproduce this issue with code similar to yours. Workaround:

```java
  db.alterTable("Accounts")
                    .alterColumn(field(name("dtCreated"), Timestamp.class)).defaultValue(DSL.currentTimestamp())
                    .execute()
```

The fix is not to inline the bind variable, but to use jOOQ's internal `Tools.field()`, which implements a guard for such "generic accidents"

**Comment by lukaseder** [comment]
This fix will be merged to 3.5.5 (#5143), 3.6.5 (#5142), and 3.7.4 (#5141)
